ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -29,28 +33,45 @@ import com.daml.platform.store.interfaces.LedgerDaoContractsReader.{
   ContractState,
   KeyState,
 }
+import com.daml.scalautil.Statement.discard
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{Await, ExecutionContext, Future}
 
 class MutableCacheBackedContractStore(
     metrics: Metrics,
     contractsReader: LedgerDaoContractsReader,
-    signalNewLedgerHead: Offset => Unit,
+    signalNewLedgerHead: SignalNewLedgerHead,
+    subscribeToContractStateEvents: SubscribeToContractStateEvents,
     private[cache] val keyCache: StateCache[GlobalKey, ContractKeyStateValue],
     private[cache] val contractsCache: StateCache[ContractId, ContractStateValue],
-)(implicit
-    executionContext: ExecutionContext
-) extends ContractStore {
+    minBackoffStreamRestart: FiniteDuration,
+)(implicit mat: Materializer, executionContext: ExecutionContext, loggingContext: LoggingContext)
+    extends ContractStore
+    with AutoCloseable {
+
   private val logger = ContextualizedLogger.get(getClass)
-  private[cache] val cacheOffset = new AtomicLong(0L)
 
-  def consumeFrom(implicit
-      loggingContext: LoggingContext
-  ): Flow[ContractStateEvent, Unit, NotUsed] =
-    Flow[ContractStateEvent]
-      .wireTap(debugEvents(_))
-      .alsoTo(Sink.foreach(updateCaches))
-      .map(updateOffsets)
+  private[cache] val cacheEventSequentialId = new AtomicLong(EventSequentialId.beforeBegin)
+  private[cache] val cacheEventOffset = new AtomicReference(Offset.beforeBegin)
```

Comment:
@tudor-da : if possible, store these two offsets as a tuple in a single AtomicReference and thereby guarantee that all reads of both of them will be consistent.

Reply:
in the spirit of defensive programming :shield:

Reply:
I've considered it as well and dropped it after some self-debate since it's adding boilerplate. Ultimately, you're right, I'll make them atomic. Maybe I'll create some small wrapper so its accesors don't need to worry with selecting the right component of the tuple.